Ep 748 simplify pub interface

reuvenpo opened the pull request:
rewritten eng_wasm_derive::pub_interface

* Enabled usage of the \#[pub_interface] annotation on `impl Struct` items
  as well as on trait items. When using the new form, the user may use whatever name they want
  for the struct. When using the trait version the name of the struct/enum implementing
  the methods was hardcoded as `Contract`, but is now customizable by passing the 
  implementor name as an argument to the macro.
  (They can basically use any type they want now)
* Added the following diagnostics to the macro invocation:
  (when any of these is detected, the user will see a useful error message)
    * The attribute was placed on an item which is not a trait or inherent impl
    * The trait or impl methods use the `self` receiver
    * Trait methods provided a default implementation (i understand that we
      wanted the users to define the implementations on the `Contract` struct)
    * Trait method with any additional attribute
    * Impl method with a \#[no_mangle] attribute
    * Impl method with visibility which isn't private or `pub`
      (we don't want them defining `pub(crate)` or `pub(in something)` because
      that can be misleading in this context)
    * check that the constructor returns no value
    * check that when applied to an impl-item, the constructor is marked as `pub`
* Fixed issue around parsing return values from contract methods where their
  amount was incorrectly calculated
* When parsing of method inputs fails at runtime, the panic message now
  mentions the type and argument that caused the failure.
* If an unknown method is called, the panic message mentions the name of
  that method.
* Using the macro no longer requires writing `use eng_wasm::*;` in the using
  module
* restructured the module layout to make it easier to follow.
* added unittests
* documented a lot of things

The ERC20 example now generates the following code:
https://gist.github.com/rust-play/d9e375f7a587725be238ced968630d28
EDIT:
https://gist.github.com/rust-play/4a1da16589b738ee2f206d351dfd4c84

reuvenpo committed:
reformatted eng-wasm/derive with rustfmt

reuvenpo committed:
reformatted eng-wasm with rustfmt

reuvenpo committed:
rewritten eng_wasm_derive::pub_interface

* Enabled usage of the \#[pub_interface] annotation on `impl Struct` items
  as well as on trait items. When using the trait version the name of the
  struct/enum implementing the methods was (and still is) hardcoded as
  `Contract`. When using the new form, the user may use whatever name they want
  for the struct. It has no side-effects.
  (They can theoretically use any type they want now)
* Added the following diagnostics to the macro invocation:
  (when any of these is detected, the user will see a useful error message)
    * The attribute was placed on an item which is not a trait or inherent impl
    * The trait or impl methods use the `self` receiver
    * Trait methods provided a default implementation (i understand that we
      wanted the users to define the implementations on the `Contract` struct)
    * Trait method with any additional attribute
    * Impl method with a \#[no_mangle] attribute
    * Impl method with visibility which isn't private or `pub`
      (we don't want them defining `pub(crate)` or `pub(in something)` because
      that can be misleading in this context)
* Fixed issue around parsing return values from contract methods where their
  amount was incorrectly calculated
* When parsing of method inputs fails at runtime, the panic message now
  mentions the type and argument that caused the failure.
* If an unknown method is called, the panic message mentions the name of
  that method.
* Using the contract no longer requires writing `use eng_wasm::*;` in the using
  module
* restructured the module layout to make it easier to follow.
* added unittests
* documented a lot of things

The ERC20 example now generates the following code:
https://gist.github.com/rust-play/d9e375f7a587725be238ced968630d28

reuvenpo commented on eng-wasm/derive/src/eth_contract.rs:
```diff
@@ -0,0 +1,154 @@
+use std::fs::File;
```
I copied the implementation of the `eth_contract` macro to this new submodule. i did not otherwise change the implementation, except a small change related to breaking changes in `syn` 1.0

reuvenpo commented on eng-wasm/derive/src/eth_contract.rs:
```diff
@@ -0,0 +1,154 @@
+        impl EthContract{
+            fn new(addr_str: /*Address*/&str) -> Self {
+                use core::str::FromStr;
+        // Ethereum Addresses need to start with `0x` so we remove the first two characters
```
I should fix the indentation here

reuvenpo commented on eng-wasm/derive/.rustfmt.toml:
```diff
@@ -0,0 +1 @@
+# override global formatting. i want clean and standard runstfmt.
```
Th project directory contains a fustfmt configuration file with a lot of unusual settings (e.g. line length of 135?). i believe we should use the standard rustfmt settings, so at least for this directory i override the current global settings. i propose to remove the project-wide rustfmt configuration and reformat the project with the standard formatter.

If we keep this, I need to add a newline here

reuvenpo commented on eng-wasm/derive/Cargo.toml:
```diff
@@ -1,22 +1,24 @@
+proc-macro = true
+
 [dependencies]
-eng-wasm = { version = "0.1.3", path = "../" }
```
Please note that this dependency is both unnecessary, and causes a future compatibility issue. I would like to create a dependency from `eng-wasm` to `eng-wasm-derive` so that we can reexport pub_interface from eng-wasm directly (and remove the requirement to depend on `eng-wasm-derive` directly. This can not happen without bumping the version of `eng-wasm` to 0.2 as it will not be compatible with older versions of `eng-wasm-derive` which depend on `eng-wasm`

reuvenpo commented on eng-wasm/derive/Cargo.toml:
```diff
@@ -1,22 +1,24 @@
-[lib]
-proc-macro = true
+[dev-dependencies]
+syn = { version = "1.0", features = ["full", "extra-traits"] }
```
I use the `extra-traits` feature for printing and comparisons of AST objects

reuvenpo commented on eng-wasm/derive/src/into_ident.rs:
```diff
@@ -0,0 +1,9 @@
+pub trait IntoIdent {
```
I think this is fairly self-explanatory, but if not i can add a docstring. i'm surprised this is not implemented upstream, so we can file a pull request with this to https://github.com/dtolnay/syn

reuvenpo commented on eng-wasm/derive/src/lib.rs:
```diff
@@ -1,328 +1,49 @@
+///
+/// I placed it here because [that's how macro_rules scoping works][macro-rules-scoping]
+/// and rust didn't like it when i put it in a module and exported it. If you figure out a
+/// better way to hide it, please file a PR.
```
or comment here :)

reuvenpo commented on eng-wasm/derive/src/pub_interface.rs:
```diff
@@ -0,0 +1,475 @@
+fn generate_deploy_function(
+    signatures: &PubInterfaceSignatures,
+    deploy_func_name: &syn::Ident,
+) -> proc_macro2::TokenStream {
```
I should reverse the order of arguments here

reuvenpo commented on eng-wasm/derive/src/pub_interface.rs:
```diff
@@ -0,0 +1,475 @@
+pub(crate) fn impl_pub_interface(item: proc_macro2::TokenStream) -> proc_macro2::TokenStream {
+    //    let item = parse_macro_input2!(item as syn::ItemTrait);
+    //    println!("SYNTAX: {:#?}", item);
+    //    item.to_token_stream()
```
I forgot to remove this

reuvenpo commented on eng-wasm/derive/src/pub_interface.rs:
```diff
@@ -0,0 +1,475 @@
+                    // which flows back to
+                    // pwasm_abi/derive/src/lib.rs :: fn generate_eth_endpoint
+                    // which dictates how return values are serialised into the Sink.
+                    // This can be 0 is the return type is () which is correct.
```
typo "if the return value is"

reuvenpo commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -0,0 +1,600 @@
+}
+
+/// The signatures collected while parsing the macro input
+#[allow(dead_code)]
```
remove this annotation

reuvenpo commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -0,0 +1,600 @@
+        .partition::<Vec<_>, _>(|item| item.is_ok());
+
+    // Thanks to the .partition() above, we know that:
+    // all `signatures` are `Result::Ok(syn::Signature)`
```
all `trait_methods`

reuvenpo commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -0,0 +1,600 @@
+        .partition::<Vec<_>, _>(|item| item.is_ok());
+
+    // Thanks to the .partition() above, we know that:
+    // all `signatures` are `Result::Ok(syn::Signature)`
```
all `impl_methods`

reuvenpo commented on eng-wasm/derive/src/lib.rs:
```diff
@@ -1,328 +1,49 @@
-#![deny(unused_extern_crates)]
 #![feature(box_patterns)]
```
This can be avoided:
https://github.com/dtolnay/syn/issues/701

moriaab commented on eng-wasm/derive/src/into_ident.rs:
```diff
@@ -0,0 +1,9 @@
+pub trait IntoIdent {
```
Nice addition!

reuvenpo replied:
I just realised this is probably why it isn't made to be as easy to use as i made it:
https://docs.rs/quote/1.0.2/quote/macro.format_ident.html#panics

moriaab commented on eng-wasm/derive/src/lib.rs:
```diff
@@ -1,333 +1,49 @@
-#![deny(unused_extern_crates)]
```
Nice cleaning of attributes and dependencies :-)

moriaab commented on eng-wasm/derive/src/pub_interface.rs:
```diff
@@ -0,0 +1,475 @@
+use super::into_ident::IntoIdent;
+use parse_signatures::PubInterfaceSignatures;
+
+const DEFAULT_IMPLEMENTOR_NAME: &str = "Contract";
```
Excellent addition of definition and work with string constants here!

reuvenpo commented on eng-wasm/derive/src/pub_interface.rs:
```diff
@@ -0,0 +1,475 @@
+                            let mut stream = eng_wasm::eng_pwasm_abi::eth::Stream::new(args);
+                            let result = <#implementor>::#method_name(#(stream.pop::<#input_types>().expect(#expectations),)*);
+                            // 32 is the size of each argument in the serialised form
+                            // The Sink.drain_to() method might resize this array is any
```
typo: The Sink.drain_to() method might resize this array **if** any

reuvenpo commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -0,0 +1,600 @@
+/// Failures that can happen while parsing the macro input.
+///
+/// This type mostly exists to pass into syn::Error as an error message.
+#[derive(Fail, Debug)]
```
use this instead of `Fail`:
https://crates.io/crates/parse-display

moriaab commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -0,0 +1,600 @@
+//! This module defines the parsing of input for the `#[pub_interface]` macro
```
parse_signatures.rs code sets high standards :) Excellent work!

moriaab commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -0,0 +1,600 @@
+        let mut implementor: syn::Type;
+        let signatures = match pub_interface_input {
+            PubInterfaceInput::ItemTrait(item_trait) => {
+                let default_implementor_name = super::DEFAULT_IMPLEMENTOR_NAME.into_ident();
```
Please pay attention: the fixed name `DEFAULT_IMPLEMENTOR_NAME` of the structure is a *defect* that you luckily revealed. Any name should be allowed. I will appreciate if you can add this fix as well.

reuvenpo replied:
fixed :)

moriaab commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -0,0 +1,600 @@
+    if errors.is_empty() {
+        Ok(trait_methods
+            .into_iter()
+            .map(|res| res.ok().unwrap())
```
If, as you realized, `ok()` is not needed here and in line 231, please remove it.

moriaab commented on eng-wasm/derive/src/pub_interface.rs:
```diff
@@ -0,0 +1,475 @@
+    if let Some(constructor_signature) = signatures
+        .signatures
+        .iter()
+        .find(|item| item.ident == CONSTRUCTOR_NAME)
```
Please add one more check about the constructor: it should return nothing.

reuvenpo replied:
I added this check, as well as a check that the constructor is marked as public when defined in an `impl`

moriaab commented on eng-wasm/derive/src/pub_interface.rs:
```diff
@@ -0,0 +1,475 @@
+                    // pwasm_abi/derive/src/lib.rs :: fn generate_eth_endpoint
+                    // which dictates how return values are serialised into the Sink.
+                    // This can be 0 is the return type is () which is correct.
+                    syn::ReturnType::Type(_, box syn::Type::Tuple(return_tuple)) => return_tuple.elems.len(),
```
Amazing catch and fix here!

reuvenpo committed:
trivial fixups and typo fixes

reuvenpo committed:
removed use of feature(box_patterns) from eng-wasm-derive

reuvenpo committed:
changed a few uses of `result.ok().unwrap()` into `result.unwrap()`

reuvenpo committed:
replaced usage of `failure::Fail` with `parse_display::Display` for implementing Display on `eng-wasm-derive::pub_interface::parse_signatures::ParseError`

reuvenpo committed:
added checks to pub_interface related to the contract constructor

* added check that the constructor returns no value
* added check that when applied to an impl-item, the constructor is marked as `pub`

reuvenpo committed:
added interpretation of the `attr` tokens passed to the pub_interface macro

* If applied to a trait, the tokens are interpreted as the type that
  implements the trait (this affects code generation, which otherwise
  assumed the implementor was named `Contract`)
* If applied to an impl, and there are extra `attr` tokens, an error is issued.

reuvenpo committed:
removed the dependency on the order of function argument evaluation in pub_interface

Also slightly improved the generated code for dispatches to contract functions that
do not take arguments.

moriaab commented on eng-wasm/derive/src/pub_interface/parse_signatures.rs:
```diff
@@ -292,11 +363,35 @@ fn check_impl_method(
+    }
+}
+
+/// This function checks if the function has a non-unit return type.
```
Good work! It is so easy to miss the distinction between empty tuple and no return type.

moriaab reviewed (approved):
This PR improves `eng-wasm-derive` crate significantly in all aspects. Well done!

reuvenpo committed:
added two basic examples to eng-wasm

reuvenpo committed:
fixed spans of type errors in inputs/outputs of secret contract methods

reuvenpo committed:
errors about private contructors in secret contracts now point to the fn signature\n\npreviously they accidentally pointed to the missing pub token

reuvenpo committed:
added example for renaming of the implementor in eng-derive

reuvenpo committed:
When custom error happen in pub_interface, the original code is still written back

Previously the code was removed in case of error, which caused weird "missing function"
errors after macro evaluation

reuvenpo committed:
Merge remote-tracking branch 'origin/develop' into EP-748-simplify-pub_interface

moriaab commented on eng-wasm/derive/examples/trait-impl-rename.rs:
```diff
@@ -0,0 +1,29 @@
+    fn expand(input: u32) -> u64;
+}
+
+impl MyContract for MyContractImplementation {
```
I would maybe add `impl MyContract` for another structure in addition, just to emphasize the point of the argument to `pub_interface`. But it is up to you.
